### ISSUE TITLE
Not able to sync with azure AD

### ISSUE THREAD

**Issue by rajeevsonkhia** [sfonteneau/AzureADConnect_Samba4#6]
Hi Simon,

When i am doing the dry run to synchronise the samba ad and azure ad it gives below error.

Traceback (most recent call last):
  File "run_sync.py", line 209, in <module>
    run_sync(force=False)
  File "run_sync.py", line 94, in run_sync
    azure.generate_all_dict()
  File "/opt/sync-azure/libsync.py", line 82, in generate_all_dict
    list_groups = self.az.list_groups()
  File "/opt/sync-azure/AADInternals_python/AADInternals.py", line 510, in list_groups
    dict_cloudanchor_sourceanchor = self.get_dict_cloudanchor_sourceanchor()
  File "/opt/sync-azure/AADInternals_python/AADInternals.py", line 490, in get_dict_cloudanchor_sourceanchor
    for entry in self.get_syncobjects(False):
  File "/opt/sync-azure/AADInternals_python/AADInternals.py", line 538, in get_syncobjects
    return  self.xml_to_result(response,command).get('b:ResultObjects',{}).get('b:AzureADSyncObject',[])
AttributeError: 'NoneType' object has no attribute 'get'

**Comment by kedar-linuxspec** [comment]
A similar issue identified for us too. A working solution until Friday last week... stopped working on Saturday giving the exact above error. Would be great if any solution can be provided to fix this

**Comment by sfonteneau** [comment]
I can't reproduce the problem here.

To help me find you can add

```
        response = self.call_adsyncapi(envelope,command,self.tenant_id,message_id)
        print(response)
        return  self.xml_to_result(response,command).get('b:ResultObjects',{}).get('b:AzureADSyncObject',[])
```

at line 538 in /opt/sync-azure/AADInternals_python/AADInternals.py

You can then run the script again and send me the result

You can also try changing:

```
def get_syncobjects(self,fullsync=True,version=2):
```

to 

```
def get_syncobjects(self,fullsync=False,version=2):
```

we'll see if that changes anything.

**Comment by sfonteneau** [comment]
in any case I will be able to modify the script to no longer use get_syncobjects.

Unfortunately, deleting groups and devices will work less well than before. This part of the script is a DIY to retrieve the sourceanchoor of the groups. The microsoft api does not allow it otherwise... strange for an attribute that microsoft considers very important...

**Comment by kedar-linuxspec** [comment]
Hey- sure.. thanks.. getting it checked. We also now got an error as below. 
---------------------------------
Traceback (most recent call last):
  File "/opt/sync-azure/run_sync.py", line 209, in <module>
    run_sync(force=False)
  File "/opt/sync-azure/run_sync.py", line 127, in run_sync
    azure.send_obj_to_az(smb.dict_all_users_samba[entry])
  File "/opt/sync-azure/libsync.py", line 56, in send_obj_to_az
    self.az.set_azureadobject(**entry)
  File "/opt/sync-azure/AADInternals_python/AADInternals.py", line 462, in set_azureadobject
    newresponse = self.xml_to_result(rawresponse,command)['b:SyncObjectResults']['b:AzureADSyncObjectResult']
  File "/opt/sync-azure/AADInternals_python/AADInternals.py", line 473, in xml_to_result
    raise Exception(dataxml["s:Envelope"]["s:Body"]['s:Fault']['s:Reason']['s:Text']['#text'])
Exception: Identity synchronization is not yet activated for this company. To activate Identity synchronization, click the Activate button on the Identity synchronization page in your Admin Portal.

I could not find any setting which I could use to manually set the synchronization to 'On' in the Admin Portal

**Comment by sfonteneau** [comment]
is it a recent install?

I have already been told about the problem and I had made a patch but obviously my patch is not good. : 

https://github.com/sfonteneau/AzureADConnect_Samba4/commit/a237636e7821ab3f2a342d7e33831f83a586688f

I enable synchonization only if the sqllite file does not exist:

https://github.com/sfonteneau/AzureADConnect_Samba4/blob/main/run_sync.py#L79

Normally this case only happens on new installations. :

**Comment by rajeevsonkhia** [comment]
> 
After adding the return response, it throws an error. which i have attached.
[sync-azure-error.txt](https://github.com/sfonteneau/AzureADConnect_Samba4/files/11782186/sync-azure-error.txt)

**Comment by sfonteneau** [comment]
the two problems are different

Are these recent installations or have they already worked ?

**Comment by kedar-linuxspec** [comment]
The one that was working stopped working on Saturday. Hence I set up a new test instance to check if there was any issue. However, the test instance could not sync as well and the above error is for the test instance that we setup on Saturday to check the issue. 

The sqllite file exists on the new tst instance that we have created and inspite of that it is giving the above error - ' Exception: Identity synchronization is not yet activated for this company. To activate Identity synchronization, click the Activate button on the Identity synchronization page in your Admin Portal.'

**Comment by kedar-linuxspec** [comment]
We have changed the alternate login id as email and not the UPN. Will that create an issue?

**Comment by sfonteneau** [comment]
no it's not related to that

I made 2 commits for both problem

For the message "Identity synchronization is not yet activated for this company, " Now I also check if the table is empty :
https://github.com/sfonteneau/AzureADConnect_Samba4/commit/c5ecd45b79f5e8a29c72b59223741719f744a75d

For message  "AttributeError: 'NoneType' object has no attribute 'get'" :
https://github.com/sfonteneau/AADInternals_python/commit/90dd2c0d7db4bea64f2a6709ca4b957f85fda0e9
It still seems surprising that ReadBackAzureADSyncObjects does not return anything...

**Comment by rajeevsonkhia** [comment]
It was throwing the old error only after doing the git pull. 
So i tried the fresh installation and it worked.
However on the portal it shows a warning message " **We detected you're using an outdated directory sync tool. We recommend you [go to the ‎Microsoft‎ Download Center](https://go.microsoft.com/fwlink/?LinkId=617037) to get the latest version of ‎Azure AD Connect‎.**"

Is there any way to resolve this.

**Comment by sfonteneau** [comment]
> So i tried the fresh installation and it worked

Yes it is the sub-module that has been modified

On the other hand, it is possible that given the message you had, the script now tries to systematically send all the groups each time.

It's the case ?

For the other message I opened another ticket. I don't know what changing the version number value entails

**Comment by sfonteneau** [comment]
https://github.com/sfonteneau/AzureADConnect_Samba4/blob/ffc337c28f28f017e1123bfb08f6afdc6c069480/CHANGELOG.md#2023-06-18

use_get_syncobjects  add in azure.conf if you have problème
